### PR TITLE
[emacs] [toplevel] Make emacs flag local to the toplevel.

### PR DESCRIPTION
**Pull request by ejgallego**
We remove the remaining emacs-specific code from the core of Coq, and
instead we put the few necessary functionality in `toplevel/`. The `-emacs` flag is purely a printing flag now.

### CODE REVIEW

File: stm/stm.ml
Comment (JasonGross) [inline stm/stm.ml]: Is `-emacs -batch` supported?  What would it mean?  If it's not, there should be an error message, right?  If it is, then this changes semantics, right?

**Comment by JasonGross** [comment]
> We also remove the experimental infoH tactic, it could be achieved by
other cleaner means these days using the STM API.

Isn't `infoH` still used by PG (or company-coq) internally?  I agree that it shouldn't be in the tactic syntax tree, but perhaps it should be in a `TACTIC EXTEND` in some plugin?

**Comment by ejgallego** [comment]
> Isn't infoH still used by PG (or company-coq) internally? I agree that it shouldn't be in the tactic syntax tree, but perhaps it should be in a TACTIC EXTEND in some plugin?

Up to @Matafou to decide on the future of it is, this PR is just a RFC.

PG-XML etc... should switch to speculative execution in the medium term.

File: stm/stm.ml
Comment (ejgallego) [inline stm/stm.ml]: AFAICT `-emacs -batch` makes no sense.

Comment (ejgallego) [inline stm/stm.ml]: In fact this test should be reduced to just `not !Flags.batch_mode`, `ide_slave` is another flag that should be removed once we tweak Stm.add.

Comment (JasonGross) [inline stm/stm.ml]: Well, `coqtop -emacs -compile` works and changes the way warnings are displayed (with tags rather than colors)... perhaps it should error instead? (It seems not very useful without the `<prompt>` tags, at least for applications I'm imagining, like the bug-minimizer.)

Comment (ejgallego) [inline stm/stm.ml]: Well, in fact if the semantics of `-emacs` is "modify the printer to add tags", indeed it could work well with compile. But that also confirms that the check in `Stm` was wrong.

**Comment by psteckler** [comment]
`infoH` is still in vanilla PG; you can still use it to auto-fill `destruct` patterns and so on.

Some vestige of it is in PG/xml, though it's not working properly. I just tried it with a `destruct`. With -emacs, coqtop sends back the pattern in <infoH> tags. With the XML protocol, the pattern is returned in a "notice" feedback; the feedback doesn't have any information to link it back to the `Add` containing the `infoH` tactical, so PG/xml has to do some guessing to use the feedback.

File: stm/stm.ml
Comment (Matafou) [inline stm/stm.ml]: PG uses a different path to deal with auto-compilation so -emacs -compile is probably useless and we can make it an error.
Unless people put -compile in the _coqProject file? That would raise an error in situations where it only altered the behaviour of coqtop.

**Comment by Matafou** [comment]
infoH is very useful in current PG. Please don't remove it. Once xml/pg is widely adopted shall we disrupt toplevel/pg and -emacs itself maybe.

**Comment by psteckler** [comment]
I will make `infoH` work in PG/xml.

**Comment by ejgallego** [comment]
Let's move it to a plugin.

**Comment by Matafou** [comment]
we should move -emacs as a whole to a plugin. Is this what you mean?

**Comment by ejgallego** [comment]
I got that idea from your comment, I am not sure that it is worth moving to a plugin, right now `-emacs` is fairly minimal. But however `infoH` could indeed be a plugin.

**Comment by ejgallego** [comment]
Anyways my goal was to make the toplevel a fully self-contained component, and we are very very close to it. Only a few global variables related to printing remain. Once we achieve that, the toplevel will fully be a separate program and whatever the maintainer deems appropriate so will be I guess.

**Comment by psteckler** [comment]
How is `infoH` handled in the XML protocol? Is that part of the toplevel?

File: stm/stm.ml
Comment (JasonGross) [inline stm/stm.ml]: I found a use case for `-emacs -compile`!  Bisecting bugs in `-emacs` ;-)  (I'm going to use it to bisect 5535.)

**Comment by Matafou** [comment]
:) good point. Let us make it survive for one more coq version.

P.

2017-05-19 21:06 GMT+02:00 Jason Gross <notifications@github.com>:

**Comment by ejgallego** [comment]
So @Matafou , is it OK to put in 8.7 the rest of the changes (leaving `infoH` as is) ?

I think so, so if you agree I'll try to schedule this for 8.7.

**Comment by ejgallego** [comment]
Ok, PR updated. @Matafou there are slight changes on printing in a couple of places, let me know if you are OK with them.

**Comment by Matafou** [comment]
Will have a look this week hopefully.
P

**Comment by ejgallego** [comment]
Merci !

**Comment by Matafou** [comment]
currently testing.

**Comment by Matafou** [comment]
I did my tests and everything seems fine concerning proofgeneral (based on toplevel).

**Review by Matafou — APPROVED** [review]
LGTM

**Comment by maximedenes** [comment]
Test suite seems broken.

**Comment by ejgallego** [comment]
Fixed, indeed the test that failed was a new test added since the first version of this PR :)

**Comment by Zimmi48** [comment]
But your fix is to change the test-suite!?

**Comment by ejgallego** [comment]
> But your fix is to change the test-suite!?

And?

**Comment by Zimmi48** [comment]
What is the purpose of this test file if we can change it at will?!

**Comment by ejgallego** [comment]
> What is the purpose of this test file if we can change it at will?!

Maybe you can explain us what is the purpose of this test and what is the rationale to keep the old version.

**Comment by ejgallego** [comment]
Indeed @Matafou I assumed from your OK that PG should be OK with removing the dummy `\n(dependent subgoals` bit , could you double check?

**Comment by Zimmi48** [comment]
I have no thing specifically against changing of Coq outputs thing and thus having to adapt test files but I just want to make sure that this is not done by accident :) Hopefully @Matafou will confirm.

**Comment by ejgallego** [comment]
Have you read what @Matafou wrote just above?

**Comment by Zimmi48** [comment]
You mean this I guess https://github.com/coq/coq/pull/647#issuecomment-305129433
Yes, but I think you can understand that the purpose of test files is precisely to catch mistakes were not caught by humans.
So even if @Matafou approved this PR before the test-suite fix, I think he should also approve it after.

**Comment by ejgallego** [comment]
Indeed @Matafou maybe we should be careful here and test that `coq-hide-additional-subgoals-switch` et al. are still working, in particular with `(setq coq-hide-additional-subgoals 't)`.

I am all for implementing a `Show Structured Goals` so goals can be properly parsed back, but that's your call.

**Comment by Zimmi48** [comment]
In fact, the more Emacs output test files we will have, the more we will be able to anticipate when a PR breaks PG, like earlier PRs of yours did at a time when there were no such test files. Given how important PG is for all of us on an every-day basis, we should really try hard not to break it carelessly.

**Comment by ejgallego** [comment]
> In fact, the more Emacs output test files we will have, the more we will be able to anticipate when a PR breaks PG, like earlier PRs of yours did at a time when there were no such test files. Given how important PG is for all of us on an every-day basis, we should really try hard not to break it carelessly.

IMHO you are missing the full picture here; currently, PG uses a protocol that is extremely fragile by design, and essentially unspecifiable and untestable.

The solution here is not to tie Coq's output to a particular concrete textual output format, but to move the interaction model between PG and Coq towards machine-friendly formats that can be parsed reliably.

**Comment by Zimmi48** [comment]
I'm not missing it. You talk about it all the time and I fully support your efforts. Don't be contemptuous just because I come here and warn you about carelessly changing a test file so that your PR passes. Deleting the whole test-suite would have worked too. It would have even improved the build time!

**Comment by ejgallego** [comment]
> I come here and warn you about carelessly changing a test file

Well, again, if you read the discussion the new output format has the explicit approval of @Matafou , so it was not done "carelessly".

There is indeed a subtle point that is to be sure we have tested the change wrt `coq-hide-subgoals`, but that realization wasn't what you pointed out. Maybe it is just me but I fail to see what:

> But your fix is to change the test-suite!?

does contribute to the discussion. I would appreciate concrete technical suggestions wrt what we are trying to do here.

**Comment by Matafou** [comment]
I think the goal of the test not to verify that pg works well. It is simply to check that nothing changed by mistake. 

This time things changed but it was not by mistake, @ejgallego made a proposition and asked for opinions. 
I agreed on the change because 1) it seems to work OK in the few tests I made. 2) If something I missed appears to be broken afterward it will be easy to fix.

I think we are precisely in a case where changing the test case is what we want.

**Comment by maximedenes** [comment]
All green: https://travis-ci.org/ejgallego/coq/builds/238351563

**Pull request merged**
